lib/store.js

```diff
@@ -0,0 +1,141 @@
+        if(err) {
+          logger.error('Mongo error when trying to find key state.', err);
+          return callback(new BedrockError(
+            'Failed to find key state due to internal error.',
+            'InternalError',
+            {namespace: namespace, httpStatusCode: 500, 'public': true}, err));
+        }
+        if(!record) {
+          return callback(new BedrockError(
+            'Key state not found.',
+            'NotFound',
+            {namespace: namespace, httpStatusCode: 404, 'public': true}));
+        }
+        callback(null, record.state);
+      });
+  }
+
+  /**
+   * Initialize a namespaced keystore.
+   *
+   * @param namespace the namespace to initialize.
+   * @param options the options to use:
+   *   sign the signing options:
+   *     algorithm the JWT signing algorithm.
+   *     tokenTtlInSecs the JWT token TTL.
+   *   [key] the key options:
+   *     id the key ID.
+   *   [verify] the verify options:
+   *     algorithms an array of acceptable HMAC signature algorithms.
+   *     clockToleranceInSecs the acceptable clock skew.
+   * @param callback(err) called once the operation completes.
+   */
+  provision(namespace, options, callback) {
+    let ns;
+    if(namespace === 'RSA') {
+      this.namespaces[namespace] = ns = new NamespaceRsa();
+    } else if(namespace === 'HMAC') {
+      this.namespaces[namespace] = ns = new NamespaceHmac();
+    } else {
+      return callback(new BedrockError(
+        'Invalid namespace specified.', 'InvalidNamespace',
+        {public: false, namespace: namespace}));
+    }
+    async.auto({
+      createState: callback => ns.createState(options, callback),
+      storeState: ['createState', (callback, results) => {
+        const id = results.createState.id;
+        const now = Date.now();
+        const record = {
+          id: database.hash(id),
+          meta: {
+            created: now,
+            updated: now
+          },
+          state: results.createState
+        };
+        database.collections.jwtKeyStore.insert(
+          record, database.writeOptions, (err, result) => {
+            if(err) {
+              if(database.isDuplicateError(err)) {
+                // namespace is already provisioned
+                return callback();
+              }
+              logger.error('Mongo error when trying to insert key state.', err);
+              return callback(new BedrockError(
+                'Failed to insert key state due to internal error.',
+                'InternalError',
+                {namespace: id, httpStatusCode: 500, 'public': true},
+                err));
+            }
+            callback(null, result.ops[0]);
+          });
+      }]
+    }, callback);
+  }
+
+  sign(options, callback) {
+    async.auto({
+      getState: callback => this.getState(options.namespace, callback),
+      sign: ['getState', (callback, results) => {
+        this.namespaces[options.namespace].sign({
```

Comment:
This should be `self.namespaces[options.algorithm]` right? One namespace instance per algorithm? I think we're still having a naming miscommunication because it seems like this design is to create N namespace instances, one per namespace in the database. What we really want is only one of each of these things. Perhaps `Namespace` should be `NamespaceHandler` or `NamespaceType` or `Algorithm`. The algorithm identifiers like `RSA` and `HMAC` each map to a single `NamespaceHandler`/`NamespaceType`/`Algorithm` instance and there are only ever two of these instances right now.